google-beta/resource_compute_snapshot.go

```diff
@@ -1,220 +1,536 @@
+// ----------------------------------------------------------------------------
+//
+//     ***     AUTO GENERATED CODE    ***    AUTO GENERATED CODE     ***
+//
+// ----------------------------------------------------------------------------
+//
+//     This file is automatically generated by Magic Modules and manual
+//     changes will be clobbered when the file is regenerated.
+//
+//     Please read more about how to change this file in
+//     .github/CONTRIBUTING.md.
+//
+// ----------------------------------------------------------------------------
+
 package google
 
 import (
 	"fmt"
 	"log"
+	"reflect"
+	"strconv"
 	"time"
 
+	"github.com/hashicorp/terraform/helper/customdiff"
 	"github.com/hashicorp/terraform/helper/schema"
-	"google.golang.org/api/compute/v1"
-	"google.golang.org/api/googleapi"
+	compute "google.golang.org/api/compute/v1"
 )
 
+func customDiffComputeSnapshotSnapshotEncryptionKeys(diff *schema.ResourceDiff, meta interface{}) error {
+	// oC -> Old Convenience Value
+	// nC -> New Convenience Value
+	// oN -> Old New Field Value
+	// nN -> New New Field Value
+	oC, nC := diff.GetChange("snapshot_encryption_key_raw")
+	oN, nN := diff.GetChange("snapshot_encryption_key.0.raw_key")
+
+	if nC != "" && nN != "" {
+		return fmt.Errorf("can't use snapshot_encryption_key_raw and snapshot_encryption_key.0.raw_key at the same time." +
+			"If you're removing snapshot_encryption_key.0.raw_key, set the value to \"\" instead. This is due to limitations in terraform plan.")
+	}
+
+	// Either field (convenience or new) has a value
+	// and then has another different value, so we ForceNew.
+	// We need to handle _EVERY_ ForceNew case in this diff
+	if oC != "" && nC != "" && oC != nC {
+		return diff.ForceNew("snapshot_encryption_key_raw")
+	}
+
+	if oN != "" && nN != "" && oN != nN {
+		return diff.ForceNew("snapshot_encryption_key.0.raw_key")
+	}
+
+	// Our resource isn't using either field, then uses one;
+	// ForceNew on whichever one is now using it.
+	if (oC == "" && oN == "" && nC != "") || (oC == "" && oN == "" && nN != "") {
+		if oC == "" && nC != "" {
+			return diff.ForceNew("snapshot_encryption_key_raw")
+		} else {
+			return diff.ForceNew("snapshot_encryption_key.0.raw_key")
+		}
+	}
+
+	// convenience no longer used
+	if oC != "" && nC == "" {
+		if nN == "" {
+			// convenience is being nulled, and the new field is empty as well
+			// we've stopped using the field altogether
+			return diff.ForceNew("snapshot_encryption_key_raw")
+		} else if oC != nN {
+			// convenience is being nulled, and the new field has a new value
+			// so we ForceNew on either field
+			return diff.ForceNew("snapshot_encryption_key_raw")
+		} else {
+			// If we reach it here, we're using the same value in the new field as we had in the convenience field
+		}
+	}
+
+	// new no longer used
+	// note that it will remain _set_ because Computed fields don't
+	// lose diffs when unset, just to a 0 value.
```

Comment:
what does `lose diffs` mean?

Reply:
Ah oops - fixed comment. If I `terraform apply` with a value set for a `Computed` & `Optional` field, and then remove that field, my set value will persist in state.